### ISSUE TITLE
workflow language supports parallel execution of agents

### ISSUE THREAD

**Issue by psschwei** [i-am-bee/beeai-labs#18]
Part of #5 

As a user, I can invoke agents in parallel when building a workflow. 

"Agents" here can mean both single agents or a workflow of multiple agents.

**Comment by akihikokuroda** [comment]
@psschwei May I take this as a low priority item for now?

**Comment by psschwei** [comment]
👍

**Comment by akihikokuroda** [comment]
Making steps (my assumption of the parallelized unit) in parallel is easy but input, output and dependencies must be designed.

- The step takes one input and generates an output and the next step name.  They are saved until they are used by the step depending on this step and match as the next step as their input.
- Each step may have one or more dependencies.  When the workflow is started, the steps without dependencies are started as the prompt of the workflow as the input.
- Steps with dependencies are executed when all the dependent steps are executed and the next step matches the step.  The number of input of the step with dependencies is the same as the number of dependencies.
- At this moment, I assume that all of dependencies must be satisfied.   One of dependencies may required.

**Comment by akihikokuroda** [comment]
For the implementation side, to support **parallel**, the workflow probably needs to be started by `asyncio.run` and most of current methods in `agent.py` and `workflow.py` become `async` and called with `await`.
